spec/src/main/asciidoc/XX-config-mapping.adoc

```diff
@@ -0,0 +1,147 @@
+= Config Mapping (or Binding)
+
+Config Mapping (or Binding) allows mapping (or binding) configuration entries to complex object types (usually user
+defined), following certain rules:
+
+- A configuration path uniquely identifies object member
+- A configuration value maps (or bind) to the object member value type
+- ...
+
+== State of the Art
+
+- https://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#features.external-config.typesafe-configuration-properties[Spring Type Safe Configuration]
+- https://www.dropwizard.io/en/latest/manual/core.html#configuration[Dropwizard POJO]
+- https://guides.micronaut.io/latest/micronaut-configuration-maven-java.html#team-configuration-with-configurationproperties[Micronaut Configuration Bean]
+- https://deltaspike.apache.org/documentation/configuration.html#Interfacebasedconfiguration[DeltaSpike Interface Based Configuration]
+- https://github.com/lightbend/config#configbeanfactory[LightBend Config Bean]
+
+Popular frameworks provide such capability in one way or another. We can observe a few trends:
+
+- Most seem to favour a `class` based approach.
+- Most seem to favour a `prefix` or `namespace` to define the starting configuration path
+- Most seem to favour a convention to construct the remaining configuration path based on the complex type member name
+(with a way to override it)
+- Automatically convert the configuration entry value to fit the complex type member
+- Support container types (`Collections`, `Maps`)
+- Support nested complex types
+
+== Class vs Interface
+
+How to represent a complex object type: a `class` or an `interface`? Or maybe both?
+
+Using a `class` has some downsides:
+
+- Requires an all args constructor or setters to populate fields (or Java Beans property descriptors)
+- Possible `IllegalAccess` errors if `private` fields are supported (without setters)
+- Heavy use of reflection to build the instance
+- Additional rules to handle initialized fields
+- Additional rules to handle primitive types (they already have a default value)
```

Comment:
If we stay with CDI beans, these are not problems any more as they are injected.

Reply:
I believe we always discussed that we should not depend on CDI, so I'm assuming that core features of the specification should work standalone.

Also, CDI does not resolve these issues. You still need to set up rules on how to handle values already initialized directly in the field. This is especially problematic with Collections or Maps. Same thing for primitive types which already have a default value implicilty.

Reply:
I think that with CDI, we should recommend injecting mapped configuration objects (rather than individual values) for all but the most trivial of cases.

Reply:
Agree. On another note, I think that any CDI integration should be handled separately. This should work without CDI.